File: 206.py
# LeetCode 206 Reverse Linked List
# Reference: https://www.youtube.com/watch?v=G0_I-ZF0S38

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        while head:
            temp = None
            temp = head
            head = head.next 
            head.next = temp
        head.tail.next = None

        return head
    
    def reverseList2(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # recursive: T O(n), S O(n)

        if not head:
            return None
        
        newHead = head
        if head.next:
            newHead = self.reverseList(head.next)
            head.next.next = head
        head.next = None

        return newHead

File: 206.java
// LeetCode 206 Reverse Linked List
// Reference: https://leetcode.com/problems/reverse-linked-list/solutions/2682085/java-0ms-100-easy-understanding/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;

        while (current != null){
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
}